src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -102,86 +80,68 @@ BlkAllocStatus AppendBlkAllocator::alloc(blk_count_t nblks, const blk_alloc_hint
 
     return BlkAllocStatus::SUCCESS;
 }
 
-BlkAllocStatus AppendBlkAllocator::alloc_on_disk(BlkId const&) { return BlkAllocStatus::SUCCESS; }
-
-BlkAllocStatus AppendBlkAllocator::mark_blk_allocated(BlkId const&) { return BlkAllocStatus::SUCCESS; }
-
-void AppendBlkAllocator::free_on_disk(BlkId const&) {}
+BlkAllocStatus AppendBlkAllocator::alloc_on_disk(BlkId const&) {
+    m_is_dirty.store(true);
+    return BlkAllocStatus::SUCCESS;
+}
 
-bool AppendBlkAllocator::is_blk_alloced_on_disk(BlkId const&, bool) const { return false; }
+BlkAllocStatus AppendBlkAllocator::alloc_on_cache(BlkId const& blkid) {
+    auto new_offset = blkid.blk_num() + blkid.blk_count();
+    auto cur_offset = m_last_append_offset.load();
+    do {
+        if (cur_offset >= new_offset) { break; } // Already allocated
+    } while (!m_last_append_offset.compare_exchange_weak(cur_offset, new_offset));
+    return BlkAllocStatus::SUCCESS;
+}
 
-//
-// cp_flush doesn't need CPGuard as it is triggered by CPMgr which already handles the reference check;
-//
-// cp_flush should not block alloc/free;
-//
 void AppendBlkAllocator::cp_flush(CP* cp) {
-    const auto idx = cp->id() % MAX_CP_COUNT;
     // check if current cp's context has dirty buffer already
-    if (m_dirty_sb[idx].is_dirty) {
-        m_sb->last_append_offset = m_dirty_sb[idx].last_append_offset;
-        m_sb->freeable_nblks = m_dirty_sb[idx].freeable_nblks;
+    if (m_is_dirty.exchange(false)) {
+        m_sb->last_append_offset = m_last_append_offset.load();
+        m_sb->freeable_nblks = m_freeable_nblks.load();
 
         // write to metablk;
         m_sb.write();
-
-        // clear this dirty buff's dirty flag;
-        clear_dirty_offset(idx);
     }
 }
 
-// updating current cp's dirty buffer context;
-void AppendBlkAllocator::set_dirty_offset(const uint8_t idx) {
-    m_dirty_sb[idx].is_dirty = true;
-
-    m_dirty_sb[idx].last_append_offset = m_last_append_offset;
-    m_dirty_sb[idx].freeable_nblks = m_freeable_nblks;
-}
-
-// clearing current cp context's dirty flag;
-void AppendBlkAllocator::clear_dirty_offset(const uint8_t idx) { m_dirty_sb[idx].is_dirty = false; }
-
 //
 // free operation does:
 // 1. book keeping "total freeable" space
 // 2. if the blk being freed happens to be last block, move last_append_offset backwards accordingly;
 //
 void AppendBlkAllocator::free(const BlkId& bid) {
-    std::unique_lock lk(m_mtx);
-    auto cur_cp = hs()->cp_mgr().cp_guard();
-    const auto n = bid.blk_count();
-    m_freeable_nblks += n;
-    if (bid.blk_num() + n == m_last_append_offset) {
-        // we are freeing the the last blk id, let's rewind.
-        m_last_append_offset -= n;
-    }
-    set_dirty_offset(cur_cp->id() % MAX_CP_COUNT);
+    auto const nblks = bid.blk_count();
+    auto exp_last_offset = bid.blk_num() + nblks;
+    auto const new_offset = m_last_append_offset - nblks;
+    if (m_last_append_offset.compare_exchange_strong(exp_last_offset, new_offset)) { return; }
```

Comment:
First off @xiaoxichen this is a super catch. Intuitively felt something wrong with that code because of anti-pattern, but went ahead with it, without giving too much thought.

We have 2 choices:
Choice 1: We can fix this with following code:
```
// If we are freeing the last block, just move the offset back
blk_num_t cur_last_offset = m_last_append_offset.load();
auto const input_last_offset = bid.blk_num() + bid.blk_count();
blk_num_t new_offset;
bool freeing_in_middle{false};
do {
    if (input_last_offset == cur_last_offset) {
        new_offset = input_last_offset;
         freeing_in_middle = false;
    } else {
         new_offset = cur_last_offset;
         freeing_in_middle = true;    
    }
} while (!m_last_append_offset.compare_exchange_weak(cur_last_offset, new_offset));

if (freeing_in_middle) {
     // Freeing something in the middle, increment the count
    m_freeable_nblks.fetch_add(bid.blk_count());
}
```

This code piece ensures all conditions for example: 
```say m_last_append_offset = 10 
Thread 1 frees, blkid=8, count=1, while inside the do while, 
Thread 2 frees  blkid=9, count=1, and moved the m_last_append_offset = 9
Now thread 1 can go back and move the m_last_append_offset = 8
```
which is the correct thing to do, except there is always one additional atomic operation.

Choice 2:
Just treat any free as holes and never try to bring back m_last_append_offset.

Let me update the PR with choice 1 and we shall discuss this with implications a bit and can always remove this code.

Reply:
will 
`auto const new_offset = bid.blk_num()` works?  it avoid the pattern of read + modify an atomic